jerry-core/ecma/operations/ecma-container-object.c

```diff
@@ -614,15 +629,216 @@ ecma_op_container_delete (ecma_value_t this_arg, /**< this argument */
+{
+  ecma_map_object_t *map_object_p = ecma_op_container_get_object (this_arg, lit_id);
+
+  if (map_object_p == NULL)
+  {
+    return ECMA_VALUE_ERROR;
+  }
+
+  return ecma_op_create_iterator_object (this_arg,
+                                         ecma_builtin_get (proto_id),
+                                         iterator_type,
+                                         type);
+} /* ecma_op_container_create_iterator */
+
+/**
+ * The %{Set, Map}IteratorPrototype% object's 'next' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 23.1.5.2.1
+ *          ECMA-262 v6, 23.2.5.2.1
+ *
+ * Note:
+ *     Returned value must be freed with ecma_free_value.
+ *
+ * @return iterator result object, if success
+ *         error - otherwise
+ */
+ecma_value_t
+ecma_op_container_iterator_next (ecma_value_t this_val, /**< this argument */
+                                 ecma_pseudo_array_type_t iterator_type) /**< type of the iterator */
+{
+  if (!ecma_is_value_object (this_val))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an object."));
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (this_val);
+  ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) obj_p;
+
+  if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_PSEUDO_ARRAY
+      && ext_obj_p->u.pseudo_array.type != iterator_type)
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an iterator."));
+  }
+
+  ecma_value_t iterated_value = ext_obj_p->u.pseudo_array.u2.iterated_value;
+
+  if (ecma_is_value_empty (iterated_value))
+  {
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  ecma_map_object_t *map_object_p = (ecma_map_object_t *) (ecma_get_object_from_value (iterated_value));
+
+  ecma_object_t *internal_obj_p = ecma_get_object_from_value (map_object_p->header.u.class_prop.u.value);
+  ecma_collection_header_t *props_p = ecma_op_object_get_property_names (internal_obj_p, ECMA_LIST_NO_OPTS);
+
+  uint32_t length = props_p->item_count;
+  uint32_t index = ext_obj_p->u.pseudo_array.u1.iterator_index;
+
+  if (JERRY_UNLIKELY (index == ECMA_ITERATOR_INDEX_LIMIT))
+  {
+    /* After the ECMA_ITERATOR_INDEX_LIMIT limit is reached the [[%Iterator%NextIndex]]
+       property is stored as an internal property */
+    ecma_string_t *prop_name_p = ecma_get_magic_string (LIT_INTERNAL_MAGIC_STRING_ITERATOR_NEXT_INDEX);
+    ecma_value_t index_value = ecma_op_object_get (obj_p, prop_name_p);
+
+    if (!ecma_is_value_undefined (index_value))
+    {
+      index = (uint32_t) (ecma_get_number_from_value (index_value) + 1);
+    }
+
+    ecma_value_t put_result = ecma_op_object_put (obj_p,
+                                                  prop_name_p,
+                                                  ecma_make_uint32_value (index),
+                                                  true);
+
+    ecma_free_value (index_value);
+
+    if (!ecma_is_value_true (put_result))
```

Comment:
Since this is an internal property, should this actually be able to throw an error/fail? I might be mistaken, but as far as I can tell internal slots should always be accessible. If that is the case, we should use something else instead of `ecma_op_object_put`.

Reply:
Internal properties behave like 'average' properties, only you cannot access it from the JS context. However, an object with an internal property can be freezed, which prevents the extension of the object, therefore IMO this operation can throw an error.

Reply:
From what I can gather from the [standard](https://www.ecma-international.org/ecma-262/6.0/#sec-object-internal-methods-and-internal-slots): 

> Internal slots are not object properties and they are not inherited.

> ... internal slots are allocated as part of the process of creating an object and may not be dynamically added to an object.

Also Object.freeze seems to only affect property keys for the data and accessor properties of an object. I think it shouldn't have any impact on internal slots.

Reply:
@dbatyai I updated the PR with a new approach, which uses lower level property management function. Therefore any kind property attribute settings cannot have an affect in this code path.